### PR TITLE
[Core] adding OutputProcess

### PR DESCRIPTION
**Pull request by philbucher**
**Description**
Long awaited feature, see the discussion in #3766, esp [this commet](https://github.com/KratosMultiphysics/Kratos/issues/3766#issuecomment-572523080)

This might not yet be a perfect solution but it is very practical one that will solve many issues. Currently we often have issues due to some processes being output-processes and some being not, this PR aims to improve these struggles

Furthermore this design is consistent with the `AnalysisStage`

FYI @sunethwarna @mpentek

### CODE REVIEW

File: kratos/python_scripts/vtk_output_process.py [hunk 1]
```diff
@@ -8,9 +8,9 @@ def Factory(settings, model):
     return VtkOutputProcess(model, settings["Parameters"])
 
 
-class VtkOutputProcess(KratosMultiphysics.Process):
+class VtkOutputProcess(KratosMultiphysics.OutputProcess):
```
Comment (rubenzorrilla) [inline kratos/python_scripts/vtk_output_process.py:11]: Shouldn't we do so with the GiD output process?

File: kratos/processes/output_process.h [hunk 2]
```diff
@@ -0,0 +1,121 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
Comment (rubenzorrilla) [inline kratos/processes/output_process.h]: Header format

File: kratos/processes/output_process.h [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    /// Assignment operator.
+    OutputProcess& operator=(OutputProcess const& rOther) = delete;
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    virtual bool IsOutputStep()
+    {
+        return true;
```
Comment (rubenzorrilla) [inline kratos/processes/output_process.h:72]: Maybe we should throw an error in here.

File: kratos/processes/output_process.h [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    virtual void PrintOutput()
+    {
+
```
Comment (rubenzorrilla) [inline kratos/processes/output_process.h]: I'd throw an error in here (for me it makes no sense to implement a derived class that doesn't implement this). Indeed I'd make this pure virtual.

File: kratos/processes/output_process.h (hunk 3 above)
Comment (philbucher) [reply kratos/processes/output_process.h:72]: I would not
Some processes can always print, seems ok to me

File: kratos/processes/output_process.h (hunk 4 above)
Comment (philbucher) [reply kratos/processes/output_process.h]: I added the error

pure virtual is not possible bcs we need prototypes for registration

File: kratos/python_scripts/vtk_output_process.py (hunk 1 above)
Comment (philbucher) [reply kratos/python_scripts/vtk_output_process.py:11]: This is only an example to ensure the usage

In fact there are more places where we should use it but I would update in separate PRs

**Review by rubenzorrilla — DISMISSED** [review]

File: kratos/processes/output_process.h [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+namespace Kratos
+{
+///@addtogroup ApplicationNameApplication
+///@{
+
+///@name Kratos Classes
+///@{
+
+/// Short class definition.
```
Comment (loumalouomega) [inline kratos/processes/output_process.h]: More detailed description?

File: kratos/python_scripts/vtk_output_process.py [hunk 6]
```diff
@@ -8,9 +8,9 @@ def Factory(settings, model):
...
     def __init__(self, model, settings):
-        KratosMultiphysics.Process.__init__(self)
+        super().__init__(self)
```
Comment (loumalouomega) [inline kratos/python_scripts/vtk_output_process.py]:
```suggestion
        super().__init__()
```
Apparently is failing, the pybind documentation says this is complicated

File: kratos/processes/output_process.h [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    ///@{
+
+    virtual bool IsOutputStep()
+    {
+        return true;
+    }
+
+    virtual void PrintOutput()
+    {
+        KRATOS_ERROR << "Calling the base OutputProcess class GetDefaultParameters. Please implement the GetDefaultParameters in your derived OutputProcess class." << std::endl;
```
Comment (sunethwarna) [inline kratos/processes/output_process.h]:
```suggestion
        KRATOS_ERROR << "Calling the base OutputProcess class PrintOutput. Please implement the PrintOutput in your derived OutputProcess class." << std::endl;
```

**Review by loumalouomega — APPROVED** [review]

**Comment by pooyan-dadvand** [comment]
@jrubiogonzalez and @KratosMultiphysics/altair  This can be the place to add the last step output and eventually the modifier in future.

**Pull request merged**